### DISCUSSION TITLE
Enumeration problem

### DISCUSSION THREAD

**Discussion by ValeryPolyakov** [rdkit/rdkit#4203] [Q&A]
**Describe the bug**
I am using the following reaction:
`[C:1][N:2]=[N+:3]=[N-:4].[H:8][C:5]#[C:6][C:7]>>[C:7][C:6]1=[C:5]([H:8])[N:2]([C:1])[NH0:3]=[NH0:4]1`
with these two reactants:
`C#CC`
`CCN=[N+]=[N-]`

The component does not enumerate (both python and RDkit) 
rxn = Chem.ReactionFromSmarts(rxn)
ps=rxn.RunReactants((r1,r2))

**To Reproduce**
** the function bellow requires to input file names**
```
def Enumerate(reactionFN, outputFN, reactant1FN, reactant2FN, alignFN = None):

	rxt = None
	
	#reading reaction smarts
	if reactionFN is not None:
		with open(reactionFN) as fp:
			rxn = fp.readline()
	else:
		print("Unable to read reaction file %s"%(reactionFN))

	rxn = Chem.ReactionFromSmarts(rxn)
	reactant1 = Chem.SDMolSupplier(reactant1FN)
	print(f'Number of first reactants: {len(reactant1)}')

	reactant2 = None #optional
	if reactant2FN is not None:
		reactant2= Chem.SDMolSupplier(reactant2FN)
		print(f'Number of second reactants: {len(reactant2)}')

	align = None
	if alignFN is not None:
		align = Chem.SDMolSupplier(alignFN)[0]
	
	products = []
	product = None

	for r1 in reactant1:
		if r1 is None:
			continue
		if reactant2 is not None:
			for r2 in reactant2:
				if r2 is None:
					continue
				ps=rxn.RunReactants((r1,r2))
				if len(ps) > 0:
					product = ps[0][0]
					#I need to do this to avoid collapsing of some atoms on top of each other
					product = Chem.MolToSmiles(product)
					product = Chem.MolFromSmiles(product)

					Chem.Compute2DCoords(product)
					if align is not None:
						Chem.GenerateDepictionMatching2DStructure(product, align)
						pass
				else:
					#print("Skipping...")
					pass

				if product is not None:
					products.append(product)
		else:
			ps=rxn.RunReactants((r1,))
			product = ps[0][0]
			Chem.Compute2DCoords(product)
		
			if product is not None:
				products.append(product)

	print("Enumerated %i products"%(len(products)))

	w = Chem.SDWriter(outputFN)	
	for m in products: w.write(m)
```

**Expected behavior**

**Screenshots**

**Configuration (please complete the following information):**
 - RDKit version: 2020.09.4
 - OS: Windows
 - Python version (if relevant): 3.8
 - Are you using conda? Yes
 - If you are using conda, which channel did you install the rdkit from? default

**Comment by greglandrum** [comment]
Hi @ValeryPolyakov, I'm going to guess that the problem here is that you've included an explicit H atom in your reaction definition. If you're going to do this, you'll need to actually have H atoms present in the molecular graph of the molecules you apply the reaction to.  You'll either need to call `AddHs()` or, if you're reading molecules from an SDF where Hs are already present, be sure to include the `removeHs=False` argument when you created your SDMolSupplier.

So for your example:
```
>>> rxn1 = rdChemReactions.ReactionFromSmarts('[C:1][N:2]=[N+:3]=[N-:4].[H:8][C:5]#[C:6][C:7]>>[C:7][C:6]1=[C:5]([H:8])[N:2]([C:1])[NH0:3]=[NH0:4]1')
>>> m1 = Chem.MolFromSmiles('C#CC')
>>> m2 = Chem.MolFromSmiles('CCN=[N+]=[N-]')
>>> rxn1.RunReactants((m2,m1))
()
>>> m1h = Chem.AddHs(m1)
>>> m2h = Chem.AddHs(m2)
>>> rxn1.RunReactants((m2h,m1h))
((<rdkit.Chem.rdchem.Mol object at 0x7fccd0529330>,),)
```

Alternatively, and probably better, you can remove the explicit H atom from your reaction definition, which gets you a reaction which works regardless of whether or not Hs are present in the molecule graph:
```
>>> rxn2 = rdChemReactions.ReactionFromSmarts('[C:1][N:2]=[N+:3]=[N-:4].[CH1:5]#[C:6][C:7]>>[C:7][C:6]1=[C:5][N:2]([C:1])[NH0:3]=[NH0:4]1')
>>> rxn2.RunReactants((m2,m1))
((<rdkit.Chem.rdchem.Mol object at 0x7fccd050b450>,),)
>>> rxn2.RunReactants((m2h,m1h))
((<rdkit.Chem.rdchem.Mol object at 0x7fccd04d3630>,),)
```

**Comment by ValeryPolyakov** [comment]
Hi Greg,

Many thanks for your reply. I agree that the second method is much cleaner. I am curious as to why does it return charged nitrogens on the triazole ring:
	rxn2 = rdkit.Chem.rdChemReactions.ReactionFromSmarts('[C:1][N:2]=[N+:3]=[N-:4].[CH1:5]#[C:6][C:7]>>[C:7][C:6]1=[C:5][N:2]([C:1])[NH0:3]=[NH0:4]1')
	product = rxn2.RunReactants((m2,m1))
	if len(product) and len(product[0]) > 0:
		product = Chem.MolToSmiles(product[0][0])
		print(product)
	else:
		print("No reaction #2")
Output:
CCN1C=C(C)[N-]=[N+]1

**Comment by greglandrum** [comment]
@ValeryPolyakov If you want the reaction to remove charge, you need to tell it do so:
`[C:1][N:2]=[N+:3]=[N-:4].[CH1:5]#[C:6][C:7]>>[C:7][C:6]1=[C:5]N:2[N+0:3]=[NH0:4]1`
(I also removed the redundant `H0` from `N:3` in the products)

**Comment by greglandrum** [comment]
*[Meta]* I moved this into `Discussions` because I am trying to stop using the Issue tracker for questions and discussions.
